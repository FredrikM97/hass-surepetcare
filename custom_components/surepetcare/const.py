"""TODO."""

DOMAIN = "surepetcare"

CONF_EMAIL = "email"
CONF_PASSWORD = "password"

TOKEN = "token"
CLIENT_DEVICE_ID = "client_device_id"
FACTORY = "surepetcare_factory"
TIMEOUT_API = 10
KEY_API = "surepetcare_api"
COORDINATOR_LIST = "coordinator_list"
COORDINATOR_DICT = "coordinator_dict"
COORDINATOR = "coordinator"
