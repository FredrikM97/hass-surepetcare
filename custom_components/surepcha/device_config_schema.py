"""Device configuration schemas for Sure Petcare devices."""

from custom_components.surepcha.const import (
    LOCATION_INSIDE,
    LOCATION_OUTSIDE,
    POLLING_SPEED,
    SCAN_INTERVAL,
)
from surepcio.enums import ProductId
from voluptuous import Optional, Range, All

DEVICE_CONFIG_SCHEMAS = {
    ProductId.DUAL_SCAN_CONNECT: {
        Optional(LOCATION_INSIDE): str,
        Optional(LOCATION_OUTSIDE): str,
    },
    ProductId.DUAL_SCAN_PET_DOOR: {
        Optional(LOCATION_INSIDE): str,
        Optional(LOCATION_OUTSIDE): str,
    },
    ProductId.PET_DOOR: {
        Optional(LOCATION_INSIDE): str,
        Optional(LOCATION_OUTSIDE): str,
    },
}

# Ensure every schema includes the polling speed range
for pid in ProductId:
    schema = DEVICE_CONFIG_SCHEMAS.setdefault(pid, {})
    schema[Optional(POLLING_SPEED, default=SCAN_INTERVAL)] = All(
        int, Range(min=5, max=86400)
    )
