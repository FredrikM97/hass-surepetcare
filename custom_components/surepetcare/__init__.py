"""TODO."""

import logging
from typing import Any, List

from surepetcare.client import SurePetcareClient
from surepetcare.household import Household
from surepetcare.enums import ProductId

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import EVENT_HOMEASSISTANT_STOP, Platform
from homeassistant.core import Event, HomeAssistant, callback
from homeassistant.helpers import device_registry as dr

from .const import (
    CLIENT_DEVICE_ID,
    COORDINATOR,
    COORDINATOR_DICT,
    DOMAIN,
    FACTORY,
    KEY_API,
    TOKEN,
)
from .coordinator import SurePetCareDeviceDataUpdateCoordinator

logger = logging.getLogger(__name__)

PLATFORMS = [Platform.BINARY_SENSOR, Platform.SENSOR]


async def async_setup_entry(
    hass: HomeAssistant,
    entry: ConfigEntry,
) -> bool:
    """TODO."""
    logger.info("async_setup_entry called for entry_id=%s", entry.entry_id)

    surepetcare_data: dict[str, Any] = {}
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = surepetcare_data

    client: SurePetcareClient = SurePetcareClient()
    # Try login with token+device_id, else fallback to email+password
    token = entry.data.get(TOKEN)
    device_id = entry.data.get(CLIENT_DEVICE_ID)
    email = entry.data.get("email")
    password = entry.data.get("password")
    if token and device_id:
        await client.login(token=token, device_id=device_id)
    elif email and password:
        await client.login(email=email, password=password)
    else:
        raise Exception("No valid credentials found in config entry data")
    surepetcare_data[FACTORY] = client

    async def on_hass_stop(event: Event) -> None:
        """Close connection when hass stops."""
        await client.close()

    # Setup listeners
    entry.async_on_unload(
        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, on_hass_stop)
    )
    try:
        households: List[Household] = await client.api(Household.get_households())
        entities = []
        for household in households:
            entities.extend(await client.api(household.get_pets()))
            entities.extend(await client.api(household.get_devices()))
            # Add each hub device to hubs list
        await client.close()
    except Exception as exc:
        await client.close()
        raise Exception("Configuration not finished") from exc

    remove_stale_devices(hass, entry, entities)
    # Setup the device coordinators
    coordinator_data = {
        KEY_API: client,
        COORDINATOR_DICT: {},
    }

    for device in entities:
        if device.product != ProductId.HUB:
            continue
        entities.remove(device)
        device_id = str(device.id)
        if device_id not in coordinator_data[COORDINATOR_DICT]:
            coordinator = SurePetCareDeviceDataUpdateCoordinator(
                hass=hass,
                config_entry=entry,
                client=client,
                device=device,
            )
            await coordinator.async_config_entry_first_refresh()

            coordinator_data[COORDINATOR_DICT][device_id] = coordinator
        else:
            logger.warning("Coordinator already exists for device %s", device_id)

    for device in entities:
        device_id = str(device.id)
        if device_id not in coordinator_data[COORDINATOR_DICT]:
            coordinator = SurePetCareDeviceDataUpdateCoordinator(
                hass=hass,
                config_entry=entry,
                client=client,
                device=device,
            )
            await coordinator.async_config_entry_first_refresh()

            coordinator_data[COORDINATOR_DICT][device_id] = coordinator
        else:
            logger.warning("Coordinator already exists for device %s", device_id)

    surepetcare_data[COORDINATOR] = coordinator_data

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
    if unload_ok:
        surepetcare_data = hass.data[DOMAIN].pop(entry.entry_id)
        client: SurePetcareClient = surepetcare_data[FACTORY]
        await client.close()

    return unload_ok


@callback
def remove_stale_devices(
    hass: HomeAssistant, config_entry: ConfigEntry, devices: list[Any]
) -> None:
    """Remove stale devices from device registry. TODO: Work in progress and not functional yet"""

    device_registry = dr.async_get(hass)
    device_entries = dr.async_entries_for_config_entry(
        device_registry, config_entry.entry_id
    )
    all_device_ids = {str(device.id) for device in devices}
    for device_entry in device_entries:
        device_id: str | None = None  # Only define here
        # Check that device part of DOMAIN
        for identifier in device_entry.identifiers:
            if identifier[0] != DOMAIN:
                continue

            _id = identifier[1]
            device_id = str(_id)
        if device_id is None or device_id not in all_device_ids:
            logger.info(
                "Removing stale device entry %s for config entry %s",
                device_entry.id,
                config_entry.entry_id,
            )
            device_registry.async_update_device(
                device_entry.id, remove_config_entry_id=config_entry.entry_id
            )
