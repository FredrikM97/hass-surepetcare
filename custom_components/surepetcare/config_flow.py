"""Config flow for SurePetCare integration."""

import logging
import types
import uuid
from typing import Any

import voluptuous as vol

from homeassistant import config_entries
from surepetcare.client import SurePetcareClient
from .const import CONF_EMAIL, CONF_PASSWORD, DOMAIN
from .device_config_schema import DEVICE_CONFIG_SCHEMAS

logger = logging.getLogger(__name__)

STEP_USER_DATA_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_EMAIL): str,
        vol.Required(CONF_PASSWORD): str,
    }
)


def add_method(instance, id):
    """Add func as a method to cls."""
    name = f"async_step_{id}"

    async def device_step(self, user_input=None):
        return await self.async_step_configure_device(user_input, id=id)

    device_step.__name__ = name
    setattr(instance, name, types.MethodType(device_step, instance))
    # return device_step  # Optional: allows normal use of func


class SurePetCareConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
    """Handle a config flow for SurePetCare integration."""

    VERSION = 1

    def __init__(self) -> None:
        """Initialize the config flow."""
        self.devices: list[Any] = []
        self.device_configs: list[dict[str, Any]] = []
        self.token: str | None = None
        self.client_device_id: str | None = None

    async def async_step_user(self, user_input: dict[str, Any] | None = None):
        """Authenticate and fetch devices."""
        errors = {}

        if user_input is not None:
            email = user_input[CONF_EMAIL]
            password = user_input[CONF_PASSWORD]
            client = SurePetcareClient()

            if not await client.login(email=email, password=password):
                errors["base"] = "auth_failed"
            else:
                self.token = getattr(client, "_token", None)
                self.client_device_id = uuid.uuid4().hex
                household_ids = [
                    household["id"] for household in (await client.get_households())
                ]
                devices = await client.get_devices(household_ids)

                if not devices:
                    return self.async_abort(reason="no_devices_found")

                self.devices = {d.id: d for d in devices}
                await client.close()

                return await self.async_step_select_device()

        return self.async_show_form(
            step_id="user",
            data_schema=STEP_USER_DATA_SCHEMA,
            errors=errors,
        )

    async def async_step_select_device(self, user_input: dict[str, Any] | None = None):
        """Let the user select which device to configure next using a menu."""
        if not self.devices:
            return self.async_abort(reason="no_devices_found")

        # Filter out already-configured devices
        configured_ids = {str(d["id"]) for d in self.device_configs}
        available_devices = [
            d for d in self.devices.values() if str(d.id) not in configured_ids
        ]
        # Always show the menu if there are devices left
        if not available_devices:
            return self.async_step_create_entry()

        # Dynamically create methods for each device to handle configuration
        # This allows us to have a separate step for each device
        # Maybe not the best design, but it keeps the flow "simple"
        for device in self.devices.values():
            add_method(self, device.id)
            # setattr(self, method_name, types.MethodType(device_step, self))

        menu_options = {str(d.id): d.name for d in available_devices}
        menu_options["create_entry"] = "Finish configuration"

        # if user_input is not None:
        #    selected_id = user_input["menu_option"]
        #    if selected_id == "skip":
        #        return self._create_entry_with_devices()
        # self._device_idx = next(
        #    idx for idx, d in enumerate(self.devices) if str(d.id) == selected_id
        # )
        # Always go to configure_device, even if only one device left
        #    return await self.async_step_configure_device()

        return self.async_show_menu(
            step_id="select_device",
            menu_options=menu_options,
            description_placeholders={},
        )

    async def async_step_configure_device(
        self, user_input: dict[str, Any] | None = None, id=None
    ):
        """Configure a single device."""
        if id is not None:
            device = self.devices[id]
        if user_input and "id" in user_input:
            device = self.devices.get(user_input["id"])
        if user_input is not None:
            config = {
                "id": str(device.id),
                "name": device.name,
                "product_id": device.product_id,
                **user_input,
            }
            self.device_configs.append(config)
            return await self.async_step_select_device()

        schema_info = DEVICE_CONFIG_SCHEMAS.get(device.product_id)
        data_schema = vol.Schema(
            {
                vol.Required("id"): device.id,
                **(
                    schema_info["schema"]
                    if schema_info and schema_info["schema"]
                    else {}
                ),
            }
        )

        description = f"Configure: {device.name}"

        return self.async_show_form(
            step_id="configure_device",
            data_schema=data_schema,
            description_placeholders={"device_name": device.name, "info": description},
        )

    async def async_step_create_entry(self, user_input: dict[str, Any] | None = None):
        """Create the config entry with all device configs."""
        return self.async_create_entry(
            title="SurePetCare Devices",
            data={
                "token": self.token,
                "client_device_id": self.client_device_id,
                "devices": self.device_configs,
            },
        )
