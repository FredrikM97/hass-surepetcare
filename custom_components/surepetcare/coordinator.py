from datetime import timedelta
import logging
from typing import Any

from surepetcare.client import SurePetcareClient
from surepetcare.devices.device import SurepyDevice

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator

logger = logging.getLogger(__name__)

SCAN_INTERVAL = 300


class SurePetCareDeviceDataUpdateCoordinator(DataUpdateCoordinator[Any]):
    """Coordinator to manage data for a specific SurePetCare device."""

    config_entry: ConfigEntry

    def __init__(
        self,
        hass: HomeAssistant,
        config_entry: ConfigEntry,
        client: SurePetcareClient,
        device: SurepyDevice,
    ) -> None:
        """Initialize device coordinator."""
        self.client = client
        self.device = device
        self._exception: Exception | None = None

        super().__init__(
            hass,
            logger,
            config_entry=config_entry,
            name=f"Update coordinator for {device}",
            update_interval=timedelta(seconds=SCAN_INTERVAL),
        )

    @callback
    def _observe_update(self, device: Any) -> None:
        """Update the coordinator for a device when a change is detected."""
        self.async_set_updated_data(data=device)

    async def _async_update_data(self) -> Any:
        """Fetch data from the api for a specific device."""
        return await self.client.api(self.device.refresh())
