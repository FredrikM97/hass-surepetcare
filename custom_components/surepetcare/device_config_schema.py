"""TODO."""

from voluptuous import Required

from surepetcare.enums import ProductId

DEVICE_CONFIG_SCHEMAS = {
    ProductId.FEEDER_CONNECT: {
        "schema": {
            Required("location_inside"): str,
            Required("location_outside"): str,
        },
    },
    ProductId.DUAL_SCAN_PET_DOOR: {
        "schema": {
            Required("location_inside"): str,
            Required("location_outside"): str,
        },
    },
    ProductId.HUB: {
        "schema": None,
    },
    ProductId.PET: {
        "schema": None,
    },
}
