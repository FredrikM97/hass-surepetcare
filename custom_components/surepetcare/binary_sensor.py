"""Support for Sure Petcare binary sensors."""

from dataclasses import dataclass
from datetime import datetime
import logging
from typing import cast

from surepcio.enums import ProductId
from surepcio.devices.device import SurePetCareBase
from surepcio.devices.dual_scan_connect import Curfew
from homeassistant.components.binary_sensor import (
    BinarySensorDeviceClass,
    BinarySensorEntity,
    BinarySensorEntityDescription,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity import EntityCategory
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .const import COORDINATOR, COORDINATOR_DICT, DOMAIN, KEY_API
from .coordinator import SurePetCareDeviceDataUpdateCoordinator
from .entity import (
    SurePetCareBaseEntity,
    SurePetCareBaseEntityDescription,
)

logger = logging.getLogger(__name__)


def _next_enabled_future_curfew(device: SurePetCareBase, r: ConfigEntry) -> bool | None:
    curfews: list[Curfew] | Curfew = device.control.curfew

    if curfews is None:
        return None

    if not isinstance(curfews, list):
        curfews = [curfews]

    now = datetime.now().time()
    return any(c.enabled and c.lock_time <= now <= c.unlock_time for c in curfews)


@dataclass(frozen=True, kw_only=True)
class SurePetCareBinarySensorEntityDescription(
    SurePetCareBaseEntityDescription, BinarySensorEntityDescription
):
    """Describes SurePetCare.binary_sensor entity."""


SENSOR_DESCRIPTIONS_AVAILABLE: tuple[SurePetCareBinarySensorEntityDescription, ...] = (
    SurePetCareBinarySensorEntityDescription(
        key="connectivity",
        translation_key="connectivity",
        device_class=BinarySensorDeviceClass.CONNECTIVITY,
        field="available",
    ),
)

SENSORS: dict[str, tuple[SurePetCareBinarySensorEntityDescription, ...]] = {
    ProductId.FEEDER_CONNECT: (
        SurePetCareBinarySensorEntityDescription(
            key="learn_mode",
            translation_key="learn_mode",
            field="status.learn_mode",
            entity_category=EntityCategory.DIAGNOSTIC,
            entity_registry_enabled_default=False,
        ),
        *SENSOR_DESCRIPTIONS_AVAILABLE,
    ),
    ProductId.DUAL_SCAN_CONNECT: (
        SurePetCareBinarySensorEntityDescription(
            key="learn_mode",
            translation_key="learn_mode",
            field="status.learn_mode",
            entity_category=EntityCategory.DIAGNOSTIC,
            entity_registry_enabled_default=False,
        ),
        SurePetCareBinarySensorEntityDescription(
            key="curfew",
            translation_key="curfew_active",
            field_fn=_next_enabled_future_curfew,
            extra_field="control.curfew",
        ),
        *SENSOR_DESCRIPTIONS_AVAILABLE,
    ),
    ProductId.PET_DOOR: (
        SurePetCareBinarySensorEntityDescription(
            key="learn_mode",
            translation_key="learn_mode",
            field="status.learn_mode",
            entity_category=EntityCategory.DIAGNOSTIC,
            entity_registry_enabled_default=False,
        ),
        SurePetCareBinarySensorEntityDescription(
            key="curfew",
            translation_key="curfew_active",
            field_fn=_next_enabled_future_curfew,
            extra_field="control.curfew",
        ),
    ),
    ProductId.DUAL_SCAN_PET_DOOR: (*SENSOR_DESCRIPTIONS_AVAILABLE,),
    ProductId.HUB: (*SENSOR_DESCRIPTIONS_AVAILABLE,),
}


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up a Surepetcare config entry."""
    coordinator_data = hass.data[DOMAIN][config_entry.entry_id][COORDINATOR]
    client = coordinator_data[KEY_API]

    entities = []
    for device_coordinator in coordinator_data[COORDINATOR_DICT].values():
        descriptions = SENSORS.get(device_coordinator.product_id, ())
        entities.extend(
            [
                SurePetCareBinarySensor(
                    device_coordinator,
                    client,
                    description=description,
                )
                for description in descriptions
            ]
        )
    async_add_entities(entities, update_before_add=True)


class SurePetCareBinarySensor(SurePetCareBaseEntity, BinarySensorEntity):
    """The platform class required by Home Assistant."""

    entity_description: SurePetCareBinarySensorEntityDescription

    def __init__(
        self,
        device_coordinator: SurePetCareDeviceDataUpdateCoordinator,
        client,
        description: SurePetCareBinarySensorEntityDescription,
    ) -> None:
        """Initialize a SurePetCare binary sensor."""
        super().__init__(
            device_coordinator=device_coordinator,
            client=client,
        )

        self.entity_description = description

        self._attr_unique_id = f"{self._attr_unique_id}-{description.key}"

    @property
    def is_on(self) -> bool | None:
        """Return true if the binary sensor is on."""
        return cast(bool, self._convert_value())
