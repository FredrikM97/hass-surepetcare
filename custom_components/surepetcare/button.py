"""Support for Sure Petcare Button."""

from dataclasses import dataclass
import logging
from typing import Any
from surepcio.enums import ProductId, HubPairMode
from surepcio import SurePetcareClient
from homeassistant.components.button import ButtonEntity, ButtonEntityDescription
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
from homeassistant.helpers.entity import EntityCategory
from custom_components.surepetcare.helper import MethodField, should_add_entity


from .const import (
    COORDINATOR,
    COORDINATOR_DICT,
    DOMAIN,
    KEY_API,
)
from .coordinator import SurePetCareDeviceDataUpdateCoordinator
from .entity import (
    SurePetCareBaseEntity,
    SurePetCareBaseEntityDescription,
)

logger = logging.getLogger(__name__)

@dataclass(frozen=True, slots=True)
class ButtonMethodField(MethodField):
    """MethodField for button-like entities, supporting on mapping."""

    on: Any = True
    def set(self, device: object, config: dict, value: Any) -> Any:
        if value is True and self.on:
            value = self.on
        return MethodField.set(self, device, config, value)



@dataclass(frozen=True, kw_only=True)
class SurePetCareButtonEntityDescription(
    SurePetCareBaseEntityDescription, ButtonEntityDescription
):
    """Describes SurePetCare button entity."""


BUTTONS: dict[str, tuple[SurePetCareButtonEntityDescription, ...]] = {    
    ProductId.HUB: (
         SurePetCareButtonEntityDescription(
            key="pairing_mode",
            translation_key="pairing_mode",
            field=ButtonMethodField(
                path="control.pairing_mode", on=HubPairMode.ON
            ),
            entity_category=EntityCategory.CONFIG,
        ),
    )
    
}


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddConfigEntryEntitiesCallback,
) -> None:
    """Set up SurePetCare sensors for each matching device."""
    coordinator_data = hass.data[DOMAIN][config_entry.entry_id][COORDINATOR]
    client = coordinator_data[KEY_API]

    entities = []
    for device_coordinator in coordinator_data[COORDINATOR_DICT].values():
        descriptions = BUTTONS.get(device_coordinator.product_id, ())
        entities.extend(
            [
                SurePetCareButton(
                    device_coordinator,
                    client,
                    description=description,
                )
                for description in descriptions
                if should_add_entity(
                    description, device_coordinator.data, config_entry.options
                )
            ]
        )
    async_add_entities(entities, update_before_add=True)


class SurePetCareButton(SurePetCareBaseEntity, ButtonEntity):
    """The platform class required by Home Assistant."""

    entity_description: SurePetCareButtonEntityDescription

    def __init__(
        self,
        device_coordinator: SurePetCareDeviceDataUpdateCoordinator,
        client: SurePetcareClient,
        description: SurePetCareButtonEntityDescription,
    ) -> None:
        """Initialize a Surepetcare sensor."""
        super().__init__(
            device_coordinator=device_coordinator,
            client=client,
        )
        self.entity_description = description
        self._attr_unique_id = f"{self._attr_unique_id}-{description.key}"

    async def async_press(self) -> None:
        """Press the button."""
        await self.send_command(1)
