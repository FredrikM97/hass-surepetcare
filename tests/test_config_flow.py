import pytest
from unittest.mock import AsyncMock, patch
from surepetcare.enums import ProductId
from homeassistant.core import HomeAssistant
from homeassistant.data_entry_flow import FlowResultType
from homeassistant import config_entries
from homeassistant.data_entry_flow import InvalidData
from custom_components.surepetcare.config_flow import DOMAIN
from homeassistant.config_entries import ConfigEntryState


@pytest.fixture
async def mock_surepy():
    with patch(
        "custom_components.surepetcare.config_flow.SurePetcareClient"
    ) as mock_client:
        instance = mock_client.return_value
        instance.login = AsyncMock(return_value=True)
        instance.get_devices = AsyncMock(
            return_value=[
                {
                    "product_id": ProductId.DUAL_SCAN_PET_DOOR.value,
                    "id": 1,
                    "name": "Flap 1",
                },
                {
                    "product_id": ProductId.FEEDER_CONNECT.value,
                    "id": 2,
                    "name": "Feeder 1",
                },
            ]
        )
        instance.token = "dummy-token"
        yield instance


async def test_flow_user_step_no_input(hass: HomeAssistant):
    """Test appropriate error when no input is provided."""
    _result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": "user"}
    )
    with pytest.raises(InvalidData):
        await hass.config_entries.flow.async_configure(
            _result["flow_id"], user_input={}
        )


async def test_user_login_failure(hass: HomeAssistant, mock_surepy):
    with patch(
        "custom_components.surepetcare.config_flow.SurePetcareClient"
    ) as mock_client:
        instance = mock_client.return_value
        instance.login = AsyncMock(return_value=False)

        result = await hass.config_entries.flow.async_init(
            DOMAIN, context={"source": config_entries.SOURCE_USER}
        )

        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            user_input={"email": "baduser@example.com", "password": "wrongpass"},
        )

        assert result["type"] == FlowResultType.FORM
        assert result["errors"] == {"base": "auth_failed"}


def test_dual_scan_pet_door_schema_valid():
    from custom_components.surepetcare.device_config_schema import DEVICE_CONFIG_SCHEMAS
    from surepetcare.enums import ProductId

    schema = DEVICE_CONFIG_SCHEMAS[ProductId.DUAL_SCAN_PET_DOOR]["schema"]
    # Valid input
    valid = {"location_inside": "Hall", "location_outside": "Garden"}
    assert schema(valid) == valid
    # Invalid input (missing location_inside)
    import voluptuous as vol

    with pytest.raises(vol.Invalid):
        schema({"location_outside": "Garden"})
    # Invalid input (missing location_outside)
    with pytest.raises(vol.Invalid):
        schema({"location_inside": "Hall"})


async def test_device_subentry_flow(hass: HomeAssistant, mock_surepy):
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": "user"}
    )
    result = await hass.config_entries.flow.async_configure(
        result["flow_id"],
        user_input={"email": "user@example.com", "password": "test1234"},
    )

    entry: config_entries.ConfigEntry = result.get("result")

    subentry_result = await hass.config_entries.subentries.async_init(
        (entry.entry_id, "device"), context={"source": "user"}
    )
    assert subentry_result["type"] == "form"
    product_id_field = subentry_result["data_schema"].schema["product_id"].container
    product_id = next(iter(product_id_field))
    subentry_result = await hass.config_entries.subentries.async_configure(
        subentry_result["flow_id"], user_input={"product_id": product_id}
    )
    assert subentry_result["type"] == "form"
    user_input = {
        "device_1": {"location_inside": "Living Room", "location_outside": "Garden"}
    }
    subentry_result = await hass.config_entries.subentries.async_configure(
        subentry_result["flow_id"], user_input=user_input
    )
    assert subentry_result["type"] == "create_entry"
    assert "devices" in subentry_result["data"]
    assert subentry_result["data"]["product_id"] == product_id
    devices = subentry_result["data"]["devices"]
    assert devices["device_1"]["location_outside"] == "Garden"
    assert devices["device_1"]["location_inside"] == "Living Room"


async def test_entities_created_for_devices(hass: HomeAssistant, mock_surepy):
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": "user"}
    )
    result = await hass.config_entries.flow.async_configure(
        result["flow_id"],
        user_input={"email": "user@example.com", "password": "test1234"},
    )

    entry: config_entries.ConfigEntry = result.get("result")

    subentry_result = await hass.config_entries.subentries.async_init(
        (entry.entry_id, "device"), context={"source": "user"}
    )
    assert subentry_result["type"] == "form"
    product_id_field = subentry_result["data_schema"].schema["product_id"].container
    product_id = next(iter(product_id_field))
    subentry_result = await hass.config_entries.subentries.async_configure(
        subentry_result["flow_id"], user_input={"product_id": product_id}
    )
    assert subentry_result["type"] == "form"
    user_input = {"1": {"location_inside": "Living Room", "location_outside": "Garden"}}
    subentry_result = await hass.config_entries.subentries.async_configure(
        subentry_result["flow_id"], user_input=user_input
    )

    entry = hass.config_entries.async_get_entry(result["result"].entry_id)
    assert entry.state == ConfigEntryState.LOADED
    await hass.async_block_till_done()

    assert hass.states.get("sensor.flap_1_location") is not None
    assert hass.states.get("sensor.feeder_1_last_fed") is not None

    flap_sensor = hass.states.get("sensor.flap_1_location")
    assert flap_sensor.state is not None  # or check for a specific value
    # assert flap_sensor.attributes["some_attribute"] == expected_value
