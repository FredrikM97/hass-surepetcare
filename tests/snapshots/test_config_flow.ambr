# serializer version: 1
# name: test_options_flow_full
  FlowResultSnapshot({
    'context': mappingproxy({
    }),
    'data': dict({
      'devices': dict({
        '1299453': dict({
          'location_inside': 'Washington DC',
          'location_outside': 'Space',
          'name': 'DualScanConnect door',
          'polling_speed': 120,
          'product_id': <ProductId.DUAL_SCAN_CONNECT: 6>,
        }),
        '269654': dict({
          'name': 'Feeder',
          'product_id': <ProductId.FEEDER_CONNECT: 4>,
        }),
        '727608': dict({
          'name': 'PetDoor',
          'product_id': <ProductId.PET_DOOR: 3>,
        }),
      }),
    }),
    'description': None,
    'description_placeholders': None,
    'flow_id': <ANY>,
    'handler': None,
    'type': <FlowResultType.CREATE_ENTRY: 'create_entry'>,
  })
# ---
# name: test_options_flow_full.1
  ConfigEntrySnapshot({
    'data': dict({
      'client_device_id': '123',
      'token': 'abc',
    }),
    'disabled_by': None,
    'discovery_keys': dict({
    }),
    'domain': 'surepetcare',
    'entry_id': <ANY>,
    'minor_version': 1,
    'options': dict({
      'devices': dict({
        '1299453': dict({
          'location_inside': 'Home',
          'location_outside': 'Away',
          'name': 'DualScanConnect door',
          'product_id': <ProductId.DUAL_SCAN_CONNECT: 6>,
        }),
        '269654': dict({
          'name': 'Feeder',
          'product_id': <ProductId.FEEDER_CONNECT: 4>,
        }),
        '727608': dict({
          'name': 'PetDoor',
          'product_id': <ProductId.PET_DOOR: 3>,
        }),
      }),
    }),
    'pref_disable_new_entities': False,
    'pref_disable_polling': False,
    'source': 'user',
    'subentries': list([
    ]),
    'title': 'Test SurePetCare entry',
    'unique_id': '12345',
    'version': 1,
  })
# ---
# name: test_reconfiguration_flow
  ConfigEntrySnapshot({
    'data': dict({
      'client_device_id': '123',
      'token': 'abc',
    }),
    'disabled_by': None,
    'discovery_keys': dict({
    }),
    'domain': 'surepetcare',
    'entry_id': <ANY>,
    'minor_version': 1,
    'options': dict({
      'devices': dict({
        '1299453': dict({
          'name': 'Katzenklappe',
          'product_id': 6,
        }),
        '222222': dict({
          'name': 'Pet with empty status',
          'product_id': 0,
        }),
        '269654': dict({
          'name': 'Maui matlåda (old) ',
          'product_id': 4,
        }),
        '295972': dict({
          'name': 'Hem-hub',
          'product_id': 1,
        }),
        '472721': dict({
          'name': 'Maui',
          'product_id': 0,
        }),
        '727608': dict({
          'name': 'Laundry Door',
          'product_id': 3,
        }),
      }),
    }),
    'pref_disable_new_entities': False,
    'pref_disable_polling': False,
    'source': 'user',
    'subentries': list([
    ]),
    'title': 'Test SurePetCare entry',
    'unique_id': '12345',
    'version': 1,
  })
# ---
# name: test_user_flow
  FlowResultSnapshot({
    'context': dict({
      'source': 'user',
    }),
    'data': dict({
      'client_device_id': 'mocked_device_id',
      'token': 'mocked_token',
    }),
    'description': None,
    'description_placeholders': None,
    'flow_id': <ANY>,
    'handler': 'surepetcare',
    'minor_version': 1,
    'options': dict({
      'devices': dict({
        '1299453': dict({
          'name': 'Katzenklappe',
          'product_id': 6,
        }),
        '222222': dict({
          'name': 'Pet with empty status',
          'product_id': 0,
        }),
        '269654': dict({
          'name': 'Maui matlåda (old) ',
          'product_id': 4,
        }),
        '295972': dict({
          'name': 'Hem-hub',
          'product_id': 1,
        }),
        '472721': dict({
          'name': 'Maui',
          'product_id': 0,
        }),
        '727608': dict({
          'name': 'Laundry Door',
          'product_id': 3,
        }),
      }),
    }),
    'result': ConfigEntrySnapshot({
      'data': dict({
        'client_device_id': 'mocked_device_id',
        'token': 'mocked_token',
      }),
      'disabled_by': None,
      'discovery_keys': dict({
      }),
      'domain': 'surepetcare',
      'entry_id': <ANY>,
      'minor_version': 1,
      'options': dict({
        'devices': dict({
          '1299453': dict({
            'name': 'Katzenklappe',
            'product_id': 6,
          }),
          '222222': dict({
            'name': 'Pet with empty status',
            'product_id': 0,
          }),
          '269654': dict({
            'name': 'Maui matlåda (old) ',
            'product_id': 4,
          }),
          '295972': dict({
            'name': 'Hem-hub',
            'product_id': 1,
          }),
          '472721': dict({
            'name': 'Maui',
            'product_id': 0,
          }),
          '727608': dict({
            'name': 'Laundry Door',
            'product_id': 3,
          }),
        }),
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'subentries': list([
      ]),
      'title': 'SurePetCare Devices',
      'unique_id': None,
      'version': 1,
    }),
    'subentries': tuple(
    ),
    'title': 'SurePetCare Devices',
    'type': <FlowResultType.CREATE_ENTRY: 'create_entry'>,
    'version': 1,
  })
# ---
